src/Container/ServiceExtensions.php

```diff
@@ -0,0 +1,172 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Inpsyde\Modularity\Container;
+
+use Psr\Container\ContainerInterface as Container;
+
+class ServiceExtensions
+{
+    private const SERVICE_TYPE_NOT_CHANGED = 1;
+    private const SERVICE_TYPE_CHANGED = 2;
+    private const SERVICE_TYPE_NOT_OBJECT = 0;
+
+    /**
+     * @var array<string, list<callable>>
+     */
+    protected $extensions = [];
+
+    /**
+     * @param string $type
+     * @return string
+     */
+    final public static function typeId(string $type): string
+    {
+        return "@instanceof<{$type}>";
+    }
+
+    /**
+     * @param string $extensionId
+     * @param callable $extender
+     * @return static
+     */
+    public function add(string $extensionId, callable $extender): ServiceExtensions
+    {
+        isset($this->extensions[$extensionId]) or $this->extensions[$extensionId] = [];
+        $this->extensions[$extensionId][] = $extender;
+
+        return $this;
+    }
+
+    /**
+     * @param string $extensionId
+     * @return bool
+     */
+    public function has(string $extensionId): bool
+    {
+        return isset($this->extensions[$extensionId]);
+    }
+
+    /**
+     * @param mixed $service
+     * @param string $id
+     * @param Container $container
+     * @return mixed
+     */
+    final public function resolve($service, string $id, Container $container)
+    {
+        $service = $this->resolveById($id, $service, $container);
+
+        return is_object($service)
+            ? $this->resolveByType(get_class($service), $service, $container)
+            : $service;
+    }
+
+    /**
+     * @param string $id
+     * @param mixed $service
+     * @param Container $container
+     * @return mixed
+     */
+    protected function resolveById(string $id, $service, Container $container)
+    {
+        foreach ($this->extensions[$id] ?? [] as $extender) {
+            $service = $extender($service, $container);
```

Comment:
When we call `resolveByType`, we check if the returned type has changed. But when resolving "by id", we do not check the returned type. Should we try to make it consistent here and/or not allow changing the type when extending by id?

Reply:
There's a difference. When you extend by ID you know the exact _class_ of the service because is 1:1, and you're not supposed to change the object. We leave this to the good intention of the developer.

When we extend by type, we might extend via interface or parent class and not be sure of what exatc type we get, because the same extending callback might be used for a hundred different services with a hundred different classes.

Immagine there's a service that implements both `Countable` and `Iterator`.

The you have a package register an extension for `@instanceof<Iterator>`, and another package register an extension for `@instanceof<Countable>`.

When modularity resolves the service, it finds both the extension callbacks and calls them in a `foreach`.

The first extension, that expects a `Iterator`, might return a different `Iterator` implementation. The extension don't know the original class (being an interface it might be called for many different classes), it just know it was an iterator, and it returns another iterator, nothing wrong about that.

However, when the second extension callback is executed, because the first returned another iterator that is _not_ countable, there's a fatal error because the second extension calls `$ervice->count()`. It was register to be exend `Countables` so it feels free to call `count()`.

It means that when we do extensions by type, and we target interfaces, even if the extension's developers do nothing "wrong", a fatal error might occur.

When we do extension by ID, the only way a fatal error occurs is that the developer purposely changes the class, knowing which the original class was.

If we don't do any check when extending by type, it means we are telling people _"do not change the original class or a fatal error might occur"_ but when extensing by interface, there might be many classes affected, so we are basically saying _"do not change the class, ever"_. And honestly, we are making extensions by type much less powerful if we we say that the class should not be ever changed.

Reply:
Kk fine for me 👍🏻